packages/client/src/links/wsLink/wsClient/index.ts

```diff
@@ -0,0 +1,396 @@
+   * Ensures all pending and active requests are properly finalized.
+   */
+  public async close() {
+    this.inactivityTimeout.stop();
+    this.reconnectManager.stop();
+
+    const requestsToAwait: Promise<void>[] = [];
+    for (const {
+      state,
+      message,
+      end,
+      callbacks,
+    } of this.requestManager.getRequests()) {
+      if (message.method === 'subscription') {
+        callbacks.complete();
+      } else if (state === 'pending') {
+        callbacks.error(
+          TRPCClientError.from(
+            new TRPCWebSocketClosedError({
+              message: 'Closed before connection was established',
+            }),
+          ),
+        );
+      } else {
+        requestsToAwait.push(end);
+      }
+    }
+
+    await Promise.all(requestsToAwait);
+
+    await this.activeConnection.close();
+    this.connectionState.next({
+      type: 'state',
+      state: 'idle',
+      error: null,
+    });
+  }
+
+  /**
+   * Method to request the server.
+   * Handles batching of requests and provides a mechanism to abort the request.
+   *
+   * @returns A function to abort the request.
+   */
+  public request({
+    op: { id, type, input, path },
+    callbacks,
+    lastEventId,
+  }: {
+    op: Pick<Operation, 'id' | 'type' | 'path' | 'input'>;
+    callbacks: TCallbacks;
+    lastEventId?: string;
+  }) {
+    const abort = this.batchSend(
+      {
+        id,
+        method: type,
+        params: {
+          input,
+          path,
+          lastEventId,
+        },
+      },
+      callbacks,
+    );
+
+    return () => {
+      if (type === 'subscription' && this.activeConnection.isOpen()) {
+        this.send({
+          id,
+          method: 'subscription.stop',
+        });
+      }
+
+      abort();
+    };
+  }
+
+  public get connection() {
+    return backwardCompatibility(this.activeConnection);
+  }
+
+  private setupWebSocketListeners(ws: WebSocket) {
+    ws.addEventListener('open', async () => {
+      if (this.lazyEnabled) {
+        this.inactivityTimeout.start();
+      }
+
+      if (this.connectionParams) {
+        ws.send(await buildConnectionMessage(this.connectionParams));
+      }
+
+      this.callbacks.onOpen?.();
+
+      this.connectionState.next({
+        type: 'state',
+        state: 'pending',
+        error: null,
+      });
+
+      const messages = this.requestManager
+        .getActiveRequests()
+        .map(({ message }) => message);
+      if (messages.length) {
+        ws.send(JSON.stringify(messages.length === 1 ? messages[0] : messages));
+      }
+    });
+
+    ws.addEventListener('message', ({ data }) => {
+      if (typeof data !== 'string' || ['PING', 'PONG'].includes(data)) return;
+
+      const incomingMessage = JSON.parse(data) as TRPCClientIncomingMessage;
+      if ('method' in incomingMessage) {
+        return this.handleIncomingRequest(incomingMessage);
+      }
+
+      return this.handleResponseMessage(incomingMessage);
```

Comment:
You should not return `Promise`s to this event listener - we're missing handlers here that can cause uncaught promise rejections and crash there server in Node.js

Reply:
Fixed